lib/integrations/awsoidc/credentialscache.go

```diff
@@ -0,0 +1,292 @@
+
+	if err := options.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err, "creating credentials cache")
+	}
+
+	initialized := make(chan struct{})
+	gotFirstCredsOrErr := make(chan struct{})
+
+	return &CredentialsCache{
+		roleARN:                 options.RoleARN,
+		integration:             options.Integration,
+		log:                     options.Log,
+		initialized:             initialized,
+		closeInitialized:        sync.OnceFunc(func() { close(initialized) }),
+		gotFirstCredsOrErr:      gotFirstCredsOrErr,
+		closeGotFirstCredsOrErr: sync.OnceFunc(func() { close(gotFirstCredsOrErr) }),
+		credsOrErr:              credsOrErr{err: errNotReady},
+		clock:                   options.Clock,
+		stsClient:               options.STSClient,
+	}, nil
+}
+
+func (cc *CredentialsCache) SetGenerateOIDCTokenFn(fn GenerateOIDCTokenFn) {
+	cc.generateOIDCTokenFn = fn
+	cc.closeInitialized()
+}
+
+// Retrieve implements [aws.CredentialsProvider] and returns the latest cached
+// credentials, or an error if no credentials have been generated yet or the
+// last generated credentials have expired.
+func (cc *CredentialsCache) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	cc.credsOrErrMu.RLock()
+	defer cc.credsOrErrMu.RUnlock()
+
+	if cc.credsOrErr.err != nil {
+		cc.log.WarnContext(ctx, "Returning error to AWS client", errorValue(cc.credsOrErr.err))
+	}
+
+	return cc.credsOrErr.creds, cc.credsOrErr.err
+}
+
+func (cc *CredentialsCache) Run(ctx context.Context) {
+	// Wait for initialized signal before running loop.
+	select {
+	case <-cc.initialized:
+	case <-ctx.Done():
+		cc.log.DebugContext(ctx, "Context canceled before initialized.")
+		return
+	}
+
+	cc.refreshIfNeeded(ctx)
+
+	ticker := cc.clock.NewTicker(refreshCheckInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.Chan():
+			cc.refreshIfNeeded(ctx)
+		case <-ctx.Done():
+			cc.log.DebugContext(ctx, "Context canceled, stopping refresh loop.")
+			return
+		}
+	}
+}
+
+func (cc *CredentialsCache) refreshIfNeeded(ctx context.Context) {
+	credsFromCache, err := cc.Retrieve(ctx)
+	if err == nil &&
+		credsFromCache.HasKeys() &&
+		cc.clock.Now().Add(refreshBeforeExpirationPeriod).Before(credsFromCache.Expires) {
+		// No need to refresh, credentials in cache are still valid for longer
+		// than refreshBeforeExpirationPeriod
+		return
+	}
+	cc.log.DebugContext(ctx, "Refreshing credentials.")
+
+	creds, err := cc.refresh(ctx)
+	if err != nil {
+		cc.log.WarnContext(ctx, "Failed to retrieve new credentials", errorValue(err))
+		now := cc.clock.Now()
+		// If we were not able to refresh, check if existing credentials in
+		// cache are still valid. If yes, just log debug, it will be retried on
+		// next interval check.
+		if credsFromCache.HasKeys() && now.Before(credsFromCache.Expires) {
+			cc.log.DebugContext(ctx, "Using existing credentials",
+				"ttl", ttlValue{now: now, expiry: credsFromCache.Expires})
```

Comment:
I think you need to pass the address of a `ttlValue` here as this struct literal is not addressable so the pointer receiver methods (`LogValue()` in this case) are not in the method set of `ttlValue`. I think this means the `LogValue()` method will not be called for this parameter to `log.DebugContext()`.

But I am wondering what the purpose of the `ttlValue` type is anyway, since it is only used here, so you may as well compute the delta directly and log that.

Reply:
`ttlValue` implements `slog.Valuer`, so the time delta calculation is lazily computed.

Reply:
`*ttlValue` (not `ttlValue`) implements `slog.Valuer`. I'm not 100% sure but I think your method will not get called because the literal `ttlValue{...}` is not addressable so the pointer receiver methods are not in its method set. For example: https://go.dev/play/p/kSrsk3FcYGp - that print's `goodbye` but not `hello`.

But also, a `Sub()` and a `Round()` is not particularly computationally expensive that it's worth doing lazily IMHO.

Reply:
Actually, addressability has nothing to do with this - the method set of a non-pointer value only includes the methods with a non-pointer receiver. It just means you need to pass `&ttlValue{...}` instead.

Reply:
Removed. Didn't seem with the hassle.